### DISCUSSION TITLE
Android App Bundle Tool

### DISCUSSION THREAD

**Discussion by fsodic** [iBotPeaches/Apktool#2572] [Ideas]
Hi iBot,
I say thanks for your tools ^_^
I was build aab from your tools.
On your tools, i just using decompile from apktool for decode codes to human read and using this [https://developer.android.com/studio/command-line/aapt2](aapt2) for compile the resource as protobuff format. And build using [https://developer.android.com/studio/command-line/bundletool](bundletool).

On here i see if your tools using **aapt** (not aapt2) and command are similar.
But i still get some bug where the music (like mp3 extension) won't mark as **doNotCompress**

I can tell more details if you interest my idea.
Just little help from you to solve this ^_^

Sorry my english bad.
Wish you read this

**Comment by iBotPeaches** [comment]
Not really following what you mean here. Apktool currently does not have native support for `aab` files.

**Reply by fsodic** [reply]
Okay, my problem solved and i can publish aab 😀 thanks for you time too take a reply here

**Comment by iBotPeaches** [comment]
Appears AAB is the future. https://android-developers.googleblog.com/2021/06/the-future-of-android-app-bundles-is.html

I don't know what this means yet. I will begin thinking. I don't think "apktool" should, but maybe a new spin up of "aabtool". I dunno. Need to research all the little parts that make up the AAB, which appear to the intermediate chunks that aapt2 spits out.

**Reply by fsodic** [reply]
> Appears AAB is the future. https://android-developers.googleblog.com/2021/06/the-future-of-android-app-bundles-is.html
> 
> I don't know what this means yet. I will begin thinking. I don't think "apktool" should, but maybe a new spin up of "aabtool". I dunno. Need to research all the little parts that make up the AAB, which appear to the intermediate chunks that aapt2 spits out.

If you want to create aabtool, i can share some information for you.

Currently i need apktool for decompile the apk and recompile using bundletool. Before using bundletool for recompile, i use aapt2 for encrypt xml (in folder `res` and `androidManifest.xml`) as pro2buff.
Next just arrange the folder like:

- manifest/androidManifest.xml (pro2buff format)
- dex/classes.dex
- res/ (every res file in pro2buff)

Pro2buff (.pb)

**Reply by kuoi2502** [reply]
has "aabtool" been finished yet :D

**Reply by fsodic** [reply]
> has "aabtool" been finished yet :D

Yes, i can build it 😁

**Reply by kuoi2502** [reply]
i found some way to edit an aab
- unzip the file
- using protobuf, appt2 to edit some .xml file
- update asset delivery
- zip module
- then build bundle and sign it
i did it but the step zip and unzip takes time. is there anyway we can edit it without unzip and zip again

**Comment by Comnir** [comment]
@fsodic @iBotPeaches I'm not sure - what is the use case for "aabtool"?

[The 'aab' format](https://developer.android.com/guide/app-bundle/app-bundle-format) is basically a zip file that contains mandatory parts (base module) and other optional parts, such as dynamic features, different languages, native libs for different architectures, etc. This parts are in a protobuf format.

An aab is used for distributing an application to Play Store. It is not directly installable. To install, [Bundletool](https://github.com/google/bundletool) can be used to convert to APKs (an APK set - a zip file with multiple apk files), The APKs can also contain a single universal apk (as described [here](https://github.com/iBotPeaches/Apktool/discussions/2524#discussioncomment-655047)). Bundletools has knows the protobuf definitions and could be used to manipulate aab files.

The device itself will still have apk files that can be "processed" with APKTool (universal apk files with splits are not supported yet #2514)

So how would/should aabtool used?

**Comment by MrIkso** [comment]
I don't understand why this discussion was created.  This file format is essentially intermediate for uploading and application to the market.  Such files are only for developers and it does not make any sense to decompile them ... The above is well written for what it is.  I propose to close this discussion.

**Reply by iBotPeaches** [reply]
Discussions are created to answer questions. I don't think its fair to vote a close a discussion as more questions and comments are being placed.

**Comment by codingwatching** [comment]
@iBotPeaches The build step is documented here:

https://developer.android.com/studio/build/building-cmdline#bundletool-build

**Comment by AndnixSH** [comment]
I don't see any benifit of this. APK format is not dying, we can still use APK site like APK Combo to download single APK and work with it, and at least able to use Split APKs Packer (SAP) tool to merge split APK into one when there is no alternatives

I have been using APK Combo for 2 years without problem, the site itself is safe to use and it's not a piracy site since you can't download paid APK. APK Combo is basically a life saver for me.

**Reply by fsodic** [reply]
We don't discussion about how to download apk from playstore or any other site

**Reply by AndnixSH** [reply]
But why work with split apk when you can download single apk or merge APK? also aab is not an installable format, but apks format is.
Anyway, there is a modified apktool which allows to decompile split APK if you are interested https://github.com/kikfox/kikfox

**Reply by fsodic** [reply]
Android App Bundle or AAB is one of 2 format that you can publish your android apps in play store.
I don't know how much developer lost their original script cause virus or something else and no backup.
We can still publish and update the game and follow the google developers rule update.

**Comment by albeci** [comment]
Has anyone of you managed to decompile an apk file to compile in aab?

**Reply by asep1994** [reply]
> 
> 
> Has anyone of you managed to decompile an apk file to compile in aab?

yes i did it,

**Reply by albeci** [reply]
would be able to explain the process and tools so that I can do it in my app?
I've been trying for a long time and I haven't had good results :(

**Reply by albeci** [reply]
can someone explain how to do this process? pleasee

**Reply by raj-paw** [reply]
Hi @fsodic  @iBotPeaches @asep1994  , I have decompile the apk file and now when I am trying to compile the apk resource files(to convert resource files into pro2buff format)  using aapt2 command. I am getting error "error: resource has invalid entry name ". Please help me to resolve this issue.


C:\IMS\Demo>aapt2 compile -o arg --dir res
error: resource 'drawable/$avd_hide_password__0' has invalid entry name '$avd_hide_password__0.
error: resource 'drawable/$avd_hide_password__1' has invalid entry name '$avd_hide_password__1.
error: resource 'drawable/$avd_hide_password__2' has invalid entry name '$avd_hide_password__2.
error: resource 'drawable/$avd_show_password__0' has invalid entry name '$avd_show_password__0.
error: resource 'drawable/$avd_show_password__1' has invalid entry name '$avd_show_password__1.
error: resource 'drawable/$avd_show_password__2' has invalid entry name '$avd_show_password__2.
error: resource 'drawable/$ic_launcher_foreground__0' has invalid entry name '$ic_launcher_foreground__0.
res\values\public.xml: error: file failed to compile.

**Comment by fsodic** [comment]
> Has anyone of you managed to decompile an apk file to compile in aab?
Yes, i can now 😁 send me mail for more detail

**Comment by dglynch** [comment]
This discussion is over two years old now, but I would like to express my interest in this feature: being able to decompile/reverse-engineer an AAB file would be useful to me. I would like to be able to run "apktool d filename.aab" and receive a directory that contains a set of human-readable XML, YML, SMALI, etc. files that represent the contents of the specified AAB file, just like I can already do with an APK file.

I don't have any particular interest in being able to make changes to and/or recompile this output into a new AAB file, but I guess some people might want to be able to do that too.

**Reply by IgorEisberg** [reply]
Then you'll have to write a separate "aabtool". AAB is an intermediate format, not a deployable app. It doesn't use the same build process and aapt2 can't do anything with it.
